watertap/property_models/activated_sludge/tests/test_asm2d_reaction.py

```diff
@@ -0,0 +1,623 @@
+        m.fs.R1.inlet.flow_vol.fix(18446 * units.m**3 / units.day)
+        m.fs.R1.inlet.temperature.fix(298.15 * units.K)
+        m.fs.R1.inlet.pressure.fix(1 * units.atm)
+        # For aerobic operation, the final spec on O2 will be on the outlet concentration
+        # This is is to account for O2 addition under aerobic operation
+        # For now, pick a reasonable positive value for initialization
+        m.fs.R1.inlet.conc_mass_comp[0, "S_O2"].fix(10 * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "S_N2"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "S_NH4"].fix(16 * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "S_NO3"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "S_PO4"].fix(3.6 * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "S_F"].fix(30 * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "S_A"].fix(20 * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "S_I"].fix(30 * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_I"].fix(25 * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_S"].fix(125 * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_H"].fix(30 * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_PAO"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_PP"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_PHA"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_AUT"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_MeOH"].fix(EPS * units.mg / units.liter)
+        m.fs.R1.inlet.conc_mass_comp[0, "X_MeP"].fix(EPS * units.mg / units.liter)
+        # No data on TSS from EXPOsan at this point
+        m.fs.R1.inlet.conc_mass_comp[0, "X_TSS"].fix(EPS * units.mg / units.liter)
+
+        # Alkalinity was givien in mg/L based on C
+        m.fs.R1.inlet.alkalinity[0].fix(60 / 12 * units.mmol / units.liter)
+
+        m.fs.R1.volume.fix(1333 * units.m**3)
+
+        return m
+
+    @pytest.mark.component
+    def test_dof(self, model):
+        assert degrees_of_freedom(model) == 0
+
+    @pytest.mark.component
+    def test_unit_consistency(self, model):
+        assert_units_consistent(model) == 0
+
+    @pytest.mark.component
+    def test_solve(self, model):
+        model.fs.R1.initialize()
+
+        # Change spec on O2 to outlet concentration to allow for O2 addition
+        model.fs.R1.inlet.conc_mass_comp[0, "S_O2"].unfix()
+        model.fs.R1.outlet.conc_mass_comp[0, "S_O2"].fix(2 * units.mg / units.liter)
+
+        solver = get_solver()
+        results = solver.solve(model, tee=True)
+        assert check_optimal_termination(results)
+
+    @pytest.mark.component
+    def test_solution(self, model):
+        # EXPOsan calculations appear to be slightly off from this implementation
```

Comment:
By "off" do you mean the small discrepancies like 13.443 vs. 13.357 (for SA) - I would think it's close enough and might be caused by cumulative numeric errors (i.e., by having slightly different kinetic stoichiometry due to rounding)

Reply:
That is what I was referring to, however it stems primarily from the comparison of the stoichiometric coefficients themselves. Before checking the predictions in an actual reactor, I did a through comparison of the stoichiometric coefficients and they all matched with the exception of the coefficients for alkalinity (which I raised with you separately). Seeing as alkalinity contributes to most of the rate expressions, any differences in the stoichiometric coefficient for alkalinity will result in differences in the rates and thus a compounding of errors.

This is almost definitely not due to rounding (unless EXPOSan is doing some significant rounding internally), as there is no rounding applies on the WaterTAP end (everything is done to machine precision), and a 1% error is far too large to be considered close.